### PR TITLE
NonceManager outbound nonces

### PR DESCRIPTION
**Pull request by RayXpub**
## Motivation

Implements the outbound nonce logic for the `NonceManager` contract. The goal of the `NonceManager` contract is to manage senders' nonces and extract that logic from the ramps.

## Solution

- Onramp and offramp are listed as authorized callers by the owner
- Contract owner sets previous onramp addresses
- Previous ramps can only be set once
- `EVM2EVMMultiOnRamp` calls `incrementOutboundNonce` which returns the new nonce
- Handles upgradability logic to call `prevOnRamp` if necessary

A shared `AuthorizedCallers` contract has been added and integrated to the following contracts:
- `MultiAggregateRateLimiter`
- `NonceManager`

Out ouf scope:
- Inbound nonce logic (next PR)

### CODE REVIEW

File: contracts/src/v0.8/ccip/interfaces/INonceManager.sol [hunk 1]
```diff
@@ -0,0 +1,6 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+interface INonceManager {
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/interfaces/INonceManager.sol:5]: suggestion: we should add some docs explaining the interface, and the function

File: contracts/src/v0.8/ccip/interfaces/IEVM2AnyMultiOnRamp.sol [hunk 2]
```diff
@@ -8,10 +8,4 @@ interface IEVM2AnyMultiOnRamp is IEVM2AnyOnRampClient {
   /// @param destChainSelector The destination chain selector
   /// @return the next sequence number to be used
   function getExpectedNextSequenceNumber(uint64 destChainSelector) external view returns (uint64);
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/interfaces/IEVM2AnyMultiOnRamp.sol]: q: do we still need the IEVM2AnyMultiOnRamp interface? I recon we can safely drop the OffRamp interface once we integ the NonceManager, how about this one? Is this used anywhere?

File: contracts/src/v0.8/ccip/test/commitStore/MultiCommitStore.t.sol [hunk 3]
```diff
@@ -5,6 +5,7 @@ import {IMultiCommitStore} from "../../interfaces/IMultiCommitStore.sol";
 import {IPriceRegistry} from "../../interfaces/IPriceRegistry.sol";
 import {IRMN} from "../../interfaces/IRMN.sol";
 
+import {AuthorizedCallers} from "../../../shared/access/AuthorizedCallers.sol";
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/test/commitStore/MultiCommitStore.t.sol]: nit: unused import?

File: contracts/src/v0.8/shared/access/AuthorizedCallers.sol [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  EnumerableSet.AddressSet internal s_authorizedCallers;
+
+  /// @param authorizedCallers the authorized callers to set
+  constructor(address[] memory authorizedCallers) {
+    _applyAuthorizedCallerUpdates(
+      AuthorizedCallerArgs({addedCallers: authorizedCallers, removedCallers: new address[](0)})
+    );
+  }
+
+  /// @return authorizedCallers Returns all callers that are authorized to call the validation functions
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/access/AuthorizedCallers.sol]: nit: comment should be updated - it is no longer scoped to only validation functions

File: contracts/src/v0.8/shared/access/AuthorizedCallers.sol [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+      if (caller == address(0)) {
+        revert ZeroAddressNotAllowed();
+      }
+
+      s_authorizedCallers.add(caller);
+      emit AuthorizedCallerAdded(caller);
+    }
+  }
+
+  modifier onlyAuthorizedCallers() {
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/access/AuthorizedCallers.sol]: thought: would it be more flexible to use an internal function? I can foresee for gas efficiency reasons we would perform this check mid-function. We could revisit this with the multi-ARL OnRamp changes

File: contracts/src/v0.8/shared/test/access/AuthorizedCallers.t.sol [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+  function setUp() public override {
+    super.setUp();
+    s_callers.push(vm.addr(1));
+    s_callers.push(vm.addr(2));
+
+    s_authorizedCallers = new AuthorizedCallers(s_callers);
+  }
+
+  function _assertAuthorizedCallersEqual(address[] memory a, address[] memory b) internal {
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/test/access/AuthorizedCallers.t.sol]: nit: isn't `assertEq(address[], address[])` already implemented by foundry?

File: contracts/src/v0.8/ccip/test/rateLimiter/MultiAggregateRateLimiter.t.sol [hunk 7]
```diff
@@ -141,162 +136,11 @@ contract MultiAggregateRateLimiter_setPriceRegistry is MultiAggregateRateLimiter
...
 
   function test_ZeroAddress_Revert() public {
-    vm.expectRevert(MultiAggregateRateLimiter.ZeroAddressNotAllowed.selector);
+    vm.expectRevert(AuthorizedCallers.ZeroAddressNotAllowed.selector);
     s_rateLimiter.setPriceRegistry(address(0));
   }
 }
 
-contract MultiAggregateRateLimiter_setAuthorizedCallers is MultiAggregateRateLimiterSetup {
-  function test_OnlyAdd_Success() public {
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/test/rateLimiter/MultiAggregateRateLimiter.t.sol:150]: suggestion: we should add some of these test cases back in AuthorizedCallers, especially the one that skips removals because a caller is not present

File: contracts/src/v0.8/ccip/NonceManager.sol [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    uint64 outbound;
+    uint64 inbound;
+  }
+
+  /// @dev previous ramps
+  mapping(uint64 chainSelector => PreviousRamps prevRamps) private s_prevRamps;
+  /// @dev The current nonces per sender used on the on/off ramps
+  mapping(uint64 chainSelector => mapping(bytes sender => Nonce nonce)) private s_nonces;
+
+  constructor(address[] memory authorizedCallers) AuthorizedCallers(authorizedCallers) {}
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/NonceManager.sol:34]: thought: I wonder if we need the authorizedCallers in the constructor, @RensR made a point that all chain contracts will typically be deployed before the lanes are setup. Though I suppose it's possible that we still deploy the lanes + all chain contracts at once, just without the chain configurations

File: contracts/src/v0.8/ccip/NonceManager.sol [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+      prevRamps.prevOnRamp = prevRampsArg.prevRamps.prevOnRamp;
+      emit PreviousOnRampUpdated(prevRampsArg.chainSelector, prevRamps.prevOnRamp);
+    }
+  }
+
+  /// @notice Gets the previous onRamp address for the given chain selector
+  /// @param chainSelector The chain selector
+  /// @return The previous onRamp address
+  function getPrevRamps(uint64 chainSelector) external view returns (PreviousRamps memory) {
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/NonceManager.sol]: nit: we can name this `getPreviousRamps` for consistency (ditto for `s_prevRamps`)

File: contracts/src/v0.8/ccip/NonceManager.sol [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+  /// @dev Struct that contains the chain selector and the previous on/off ramps, same as PreviousRamps but with the chain selector
+  /// so that an array of these can be passed to the applyPreviousRampsUpdates function
+  struct PreviousRampsArgs {
+    uint64 chainSelector;
+    PreviousRamps prevRamps;
+  }
+
+  /// @dev Struct that contains a sender's outbound and inbound nonces
+  struct Nonce {
+    uint64 outbound;
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/NonceManager.sol]: nit: missing packing comments

File: contracts/src/v0.8/ccip/NonceManager.sol [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    uint64 inbound;
+  }
+
+  /// @dev previous ramps
+  mapping(uint64 chainSelector => PreviousRamps prevRamps) private s_prevRamps;
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/NonceManager.sol]: nit: should we name `chainSelector` as `remoteChainSelector`? (inline with the Multi-ARL namings)

File: contracts/src/v0.8/ccip/NonceManager.sol [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+  /// so that an array of these can be passed to the applyPreviousRampsUpdates function
+  struct PreviousRampsArgs {
+    uint64 chainSelector;
+    PreviousRamps prevRamps;
+  }
+
+  /// @dev Struct that contains a sender's outbound and inbound nonces
+  struct Nonce {
+    uint64 outbound;
+    uint64 inbound;
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/NonceManager.sol]: nit: should we name this with the full name, i.e. `inboundNonce`

File: contracts/src/v0.8/ccip/NonceManager.sol [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+  /// @return The new outbound nonce
+  function incrementOutboundNonce(
+    uint64 destChainSelector,
+    bytes calldata sender
+  ) external onlyAuthorizedCallers returns (uint64) {
+    Nonce storage nonce = s_nonces[destChainSelector][sender];
+
+    uint64 outboundNonce = nonce.outbound + 1;
+
+    if (outboundNonce == 0) {
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/NonceManager.sol]: issue: this can never be zero because we increment above - we need a unit test to cover this as well

File: contracts/src/v0.8/ccip/NonceManager.sol [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+
+  /// @dev Struct that contains the chain selector and the previous on/off ramps, same as PreviousRamps but with the chain selector
+  /// so that an array of these can be passed to the applyPreviousRampsUpdates function
+  struct PreviousRampsArgs {
+    uint64 chainSelector;
+    PreviousRamps prevRamps;
+  }
+
+  /// @dev Struct that contains a sender's outbound and inbound nonces
+  struct Nonce {
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/NonceManager.sol]: suggestion: there is some inconsistency between naming, since in the `IMessageValidator` and `MultiAggregateRateLimiter` we use `incoming` and `outgoing` instead of `inbound` and `outbound`.

Personally I don't have a strong preference, inbound / outbound looks good - we should apply this to the multi-ARL naming as well

File: contracts/src/v0.8/ccip/NonceManager.sol [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+
+  /// @notice Increments the outbound nonce for the given sender on the given destination chain
+  /// @param destChainSelector The destination chain selector
+  /// @param sender The encoded sender address
+  /// @return The new outbound nonce
+  function incrementOutboundNonce(
+    uint64 destChainSelector,
+    bytes calldata sender
+  ) external onlyAuthorizedCallers returns (uint64) {
+    Nonce storage nonce = s_nonces[destChainSelector][sender];
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/NonceManager.sol]: suggestion: we can re-use the `getOutboundNonce` function (either split as internal or public)

File: contracts/src/v0.8/ccip/NonceManager.sol [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+  /// @param prevRampsArgs The previous on/off ramps addresses
+  function applyPreviousRampsUpdates(PreviousRampsArgs[] calldata prevRampsArgs) external onlyOwner {
+    for (uint256 i = 0; i < prevRampsArgs.length; i++) {
+      PreviousRampsArgs calldata prevRampsArg = prevRampsArgs[i];
+
+      PreviousRamps storage prevRamps = s_prevRamps[prevRampsArg.chainSelector];
+
+      // If the previous onRamp is already set then it should not be updated
+      if (prevRamps.prevOnRamp != address(0)) {
+        revert InvalidRampUpdate();
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/NonceManager.sol]: nit: should we name it to be more specific, i.e. `PreviousRampAlreadySet()`

File: contracts/src/v0.8/ccip/test/NonceManager.t.sol [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+
+    assertEq(s_nonceManager.getOutboundNonce(DEST_CHAIN_SELECTOR, sender), 0);
+
+    uint64 outboundNonce = s_nonceManager.incrementOutboundNonce(DEST_CHAIN_SELECTOR, sender);
+    assertEq(outboundNonce, 1);
+  }
+}
+
+contract NonceManager_applyPreviousRampsUpdates is EVM2EVMMultiOnRampSetup {
+  function test_mpsUpdatess_Success() public {
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/test/NonceManager.t.sol]: nit: test naming

File: contracts/src/v0.8/ccip/test/NonceManager.t.sol [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+    address prevOnRamp = vm.addr(1);
+    NonceManager.PreviousRampsArgs[] memory previousRamps = new NonceManager.PreviousRampsArgs[](1);
+    previousRamps[0] = NonceManager.PreviousRampsArgs(DEST_CHAIN_SELECTOR, NonceManager.PreviousRamps(prevOnRamp));
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/test/NonceManager.t.sol:31]: suggestion: should add tests for 0 ramp and multiple ramp updates

File: contracts/src/v0.8/ccip/interfaces/INonceManager.sol (hunk 1 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/interfaces/INonceManager.sol:5]: I moved the comments to the interface.

File: contracts/src/v0.8/ccip/interfaces/IEVM2AnyMultiOnRamp.sol (hunk 2 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/interfaces/IEVM2AnyMultiOnRamp.sol]: Good call, I removed it.

File: contracts/src/v0.8/ccip/test/commitStore/MultiCommitStore.t.sol (hunk 3 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/test/commitStore/MultiCommitStore.t.sol]: File has been removed after latest rebase.

File: contracts/src/v0.8/shared/access/AuthorizedCallers.sol (hunk 4 above)
Comment (RayXpub) [reply contracts/src/v0.8/shared/access/AuthorizedCallers.sol]: Updated

File: contracts/src/v0.8/shared/access/AuthorizedCallers.sol (hunk 5 above)
Comment (RayXpub) [reply contracts/src/v0.8/shared/access/AuthorizedCallers.sol]: I added an internal function that is now called within the modifier so that we have both options:
https://github.com/smartcontractkit/ccip/blob/e26573e4a2a940b020e9e372bf9e958a73f1f099/contracts/src/v0.8/shared/access/AuthorizedCallers.sol#L71-L75

File: contracts/src/v0.8/shared/test/access/AuthorizedCallers.t.sol (hunk 6 above)
Comment (RayXpub) [reply contracts/src/v0.8/shared/test/access/AuthorizedCallers.t.sol]: Updated

File: contracts/src/v0.8/ccip/test/rateLimiter/MultiAggregateRateLimiter.t.sol (hunk 7 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/test/rateLimiter/MultiAggregateRateLimiter.t.sol:150]: I added them back in the `AuthorizedCallers.t.sol`.

File: contracts/src/v0.8/ccip/NonceManager.sol (hunk 8 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/NonceManager.sol:34]: No strong opinion on this one. Maybe we can revisit when we try to update the `PriceRegistry`?

File: contracts/src/v0.8/ccip/NonceManager.sol (hunk 9 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/NonceManager.sol]: Renamed

File: contracts/src/v0.8/ccip/NonceManager.sol (hunk 11 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/NonceManager.sol]: Renamed

File: contracts/src/v0.8/ccip/NonceManager.sol (hunk 12 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/NonceManager.sol]: Renamed

File: contracts/src/v0.8/ccip/NonceManager.sol (hunk 13 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/NonceManager.sol]: Good catch, updated.

File: contracts/src/v0.8/ccip/NonceManager.sol (hunk 15 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/NonceManager.sol]: Good call, updated.

File: contracts/src/v0.8/ccip/NonceManager.sol (hunk 16 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/NonceManager.sol]: Renamed

File: contracts/src/v0.8/ccip/test/NonceManager.t.sol (hunk 17 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/test/NonceManager.t.sol]: Fixed

File: contracts/src/v0.8/ccip/test/NonceManager.t.sol (hunk 18 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/test/NonceManager.t.sol:31]: Added

File: contracts/src/v0.8/ccip/NonceManager.sol (hunk 10 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/NonceManager.sol]: Added

File: contracts/src/v0.8/ccip/NonceManager.sol (hunk 8 above)
Comment (elatoskinas) [reply contracts/src/v0.8/ccip/NonceManager.sol:34]: Sounds good to revisit this across all contracts. We should create a ticket for it

File: contracts/src/v0.8/ccip/interfaces/INonceManager.sol [hunk 19]
```diff
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+/// @notice Application contracts that intend to receive messages from
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/interfaces/INonceManager.sol]: nit: the OnRamp also receives messages from the Router, but it is not expected to implement this interface - should we rephrase it to something like

> Contract interface that allows managing sender nonces

File: contracts/src/v0.8/shared/access/AuthorizedCallers.sol [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+      if (caller == address(0)) {
+        revert ZeroAddressNotAllowed();
+      }
+
+      s_authorizedCallers.add(caller);
+      emit AuthorizedCallerAdded(caller);
+    }
+  }
+
+  /// @notice validate access
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/access/AuthorizedCallers.sol]: nit: we can use the same comment

> Checks the sender and reverts if it is anyone other than a listed authorized caller

File: contracts/src/v0.8/ccip/NonceManager.sol [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+  /// @dev Struct that contains a sender's outbound and inbound nonces
+  struct Nonce {
+    uint64 outboundNonce; // ──╮ Outbound nonce used by the onramp
+    uint64 inboundNonce; // ───╯ Inbound nonce used by the offramp
+  }
+
+  /// @dev previous ramps
+  mapping(uint64 chainSelector => PreviousRamps previousRamps) private s_previousRamps;
+  /// @dev The current nonces per sender used on the on/off ramps
+  mapping(uint64 remotChainSelector => mapping(bytes sender => Nonce nonce)) private s_nonces;
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/NonceManager.sol]:
```suggestion
  mapping(uint64 remoteChainSelector => mapping(bytes sender => Nonce nonce)) private s_nonces;
```

File: contracts/src/v0.8/ccip/test/NonceManager.t.sol [hunk 22]
```diff
@@ -0,0 +1,227 @@
...
+import {Internal} from "../libraries/Internal.sol";
+import {Pool} from "../libraries/Pool.sol";
+import {RateLimiter} from "../libraries/RateLimiter.sol";
+import {EVM2EVMMultiOnRamp} from "../onRamp/EVM2EVMMultiOnRamp.sol";
+import {EVM2EVMOnRamp} from "../onRamp/EVM2EVMOnRamp.sol";
+import {EVM2EVMMultiOnRampHelper} from "./helpers/EVM2EVMMultiOnRampHelper.sol";
+import {EVM2EVMOnRampHelper} from "./helpers/EVM2EVMOnRampHelper.sol";
+import {EVM2EVMMultiOnRampSetup} from "./onRamp/EVM2EVMMultiOnRampSetup.t.sol";
+
+contract NonceManagerTest_incrementOutboundNonce is EVM2EVMMultiOnRampSetup {
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/test/NonceManager.t.sol]: suggestion: we could add another isolated unit test where we `mockCall` the prevOffRamp and use that nonce when the nonce is 0

File: contracts/src/v0.8/ccip/NonceManager.sol (hunk 14 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/NonceManager.sol]: I created another ticket for this, I'll update the `MultiAggregateRateLimiter` in another PR.

File: contracts/src/v0.8/ccip/NonceManager.sol (hunk 21 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/NonceManager.sol]: Fixed

File: contracts/src/v0.8/shared/access/AuthorizedCallers.sol (hunk 20 above)
Comment (RayXpub) [reply contracts/src/v0.8/shared/access/AuthorizedCallers.sol]: Updated

File: contracts/src/v0.8/ccip/interfaces/INonceManager.sol (hunk 19 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/interfaces/INonceManager.sol]: Updated

File: contracts/gas-snapshots/functions.gas-snapshot [hunk 23]
```diff
@@ -30,7 +30,7 @@ FunctionsBilling_UpdateConfig:test_UpdateConfig_Success() (gas: 53173)
 FunctionsBilling__DisperseFeePool:test__DisperseFeePool_RevertIfNotSet() (gas: 8810)
 FunctionsBilling__FulfillAndBill:test__FulfillAndBill_RevertIfInvalidCommitment() (gas: 13375)
 FunctionsBilling__FulfillAndBill:test__FulfillAndBill_Success() (gas: 186305)
-FunctionsBilling__StartBilling:test__FulfillAndBill_HasUniqueGlobalRequestId() (gas: 524688)
```
Comment (RensR) [inline contracts/gas-snapshots/functions.gas-snapshot]: How did this happen?

File: contracts/gas-snapshots/shared.gas-snapshot [hunk 24]
```diff
@@ -1,3 +1,12 @@
+AuthorizedCallers_applyAuthorizedCallerUpdates:test_AddAndRemove_Success() (gas: 125205)
+AuthorizedCallers_applyAuthorizedCallerUpdates:test_OnlyAdd_Success() (gas: 133100)
+AuthorizedCallers_applyAuthorizedCallerUpdates:test_OnlyCallableByOwner_Revert() (gas: 12350)
+AuthorizedCallers_applyAuthorizedCallerUpdates:test_OnlyRemove_Success() (gas: 45064)
+AuthorizedCallers_applyAuthorizedCallerUpdates:test_RemoveThenAdd_Success() (gas: 57241)
+AuthorizedCallers_applyAuthorizedCallerUpdates:test_SkipRemove_Success() (gas: 32121)
+AuthorizedCallers_applyAuthorizedCallerUpdates:test_ZeroAddressNotAllowed_Revert() (gas: 64473)
+AuthorizedCallers_constructor:test_ZeroAddressNotAllowed_Revert() (gas: 64473)
+AuthorizedCallers_constructor:test_constructor_Success() (gas: 720513)
```
Comment (RensR) [inline contracts/gas-snapshots/shared.gas-snapshot:9]: Shared changes should happen in `shared` and be ported back to ccip

File: contracts/src/v0.8/shared/access/AuthorizedCallers.sol [hunk 25]
```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.0;
+
+import {EnumerableSet} from "../../vendor/openzeppelin-solidity/v4.7.3/contracts/utils/structs/EnumerableSet.sol";
```
Comment (RensR) [inline contracts/src/v0.8/shared/access/AuthorizedCallers.sol]: ordering of imports, vendor goes at the bottom

File: contracts/src/v0.8/shared/access/AuthorizedCallers.sol [hunk 26]
```diff
@@ -0,0 +1,82 @@
...
+import {OwnerIsCreator} from "./OwnerIsCreator.sol";
+
+/// @title The AuthorizedCallers contract
+/// @notice A contract that manages multiple authorized callers
```
Comment (RensR) [inline contracts/src/v0.8/shared/access/AuthorizedCallers.sol]: Could you add a bit more info here, anything in shared requires more info imo

File: contracts/src/v0.8/shared/access/AuthorizedCallers.sol [hunk 27]
```diff
@@ -0,0 +1,82 @@
...
+  }
+
+  /// @notice Updates the list of authorized callers
+  /// @param authorizedCallerArgs Callers to add and remove
+  function applyAuthorizedCallerUpdates(AuthorizedCallerArgs memory authorizedCallerArgs) external onlyOwner {
+    _applyAuthorizedCallerUpdates(authorizedCallerArgs);
+  }
+
+  /// @notice Updates the list of authorized callers
+  /// @param authorizedCallerArgs Callers to add and remove
```
Comment (RensR) [inline contracts/src/v0.8/shared/access/AuthorizedCallers.sol]: Let's have the comment explain we do removes before adds

File: contracts/src/v0.8/shared/test/access/AuthorizedCallers.t.sol [hunk 28]
```diff
@@ -0,0 +1,186 @@
...
+import {BaseTest} from "../BaseTest.t.sol";
+
+contract AuthorizedCallers_setup is BaseTest {
+  address[] s_callers;
+
+  AuthorizedCallers s_authorizedCallers;
+
+  function setUp() public override {
+    super.setUp();
+    s_callers.push(vm.addr(1));
```
Comment (RensR) [inline contracts/src/v0.8/shared/test/access/AuthorizedCallers.t.sol]: prefer makeAddr over address(number)

File: contracts/src/v0.8/ccip/NonceManager.sol (hunk 14 above)
Comment (RensR) [reply contracts/src/v0.8/ccip/NonceManager.sol]: Since you're already changing the ARL here, why not this PR?

File: contracts/src/v0.8/ccip/NonceManager.sol [hunk 29]
```diff
@@ -0,0 +1,102 @@
...
+
+  /// @dev previous ramps
+  mapping(uint64 chainSelector => PreviousRamps previousRamps) private s_previousRamps;
+  /// @dev The current nonces per sender used on the on/off ramps
+  mapping(uint64 remoteChainSelector => mapping(bytes sender => Nonce nonce)) private s_nonces;
+
+  constructor(address[] memory authorizedCallers) AuthorizedCallers(authorizedCallers) {}
+
+  /// @inheritdoc INonceManager
+  function incrementOutboundNonce(
```
Comment (RensR) [inline contracts/src/v0.8/ccip/NonceManager.sol]: nit: rename to something that more resembles the goal, which is getting and then incrementing the nonce

File: contracts/src/v0.8/ccip/NonceManager.sol [hunk 30]
```diff
@@ -0,0 +1,102 @@
...
+    uint64 destChainSelector,
+    bytes calldata sender
```
Comment (RensR) [inline contracts/src/v0.8/ccip/NonceManager.sol]: This can be an address, as the outbound will always be EVM for this contract. 
```
  mapping(uint64 remoteChainSelector => mapping(bytes sender => Nonce nonce)) private s_nonces;
```

could just be uint64 -> address -> nonce

File: contracts/src/v0.8/ccip/NonceManager.sol [hunk 31]
```diff
@@ -0,0 +1,102 @@
...
+
+    return outboundNonce;
+  }
+
+  /// TODO: add getInboundNonce
+
+  /// @notice Updates the previous ramps addresses
+  /// @param previousRampsArgs The previous on/off ramps addresses
+  function applyPreviousRampsUpdates(PreviousRampsArgs[] calldata previousRampsArgs) external onlyOwner {
+    for (uint256 i = 0; i < previousRampsArgs.length; i++) {
```
Comment (RensR) [inline contracts/src/v0.8/ccip/NonceManager.sol]: nit ++i

File: contracts/src/v0.8/ccip/NonceManager.sol (hunk 30 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/NonceManager.sol]: > the outbound will always be EVM for this contract.

Yes but this mapping handles both inbound and outbound nonces:https://github.com/smartcontractkit/ccip/blob/a8554f83a514af46c1b8fa23d56095cd0ae3a501/contracts/src/v0.8/ccip/NonceManager.sol#L30-L33

File: contracts/src/v0.8/ccip/NonceManager.sol (hunk 31 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/NonceManager.sol]: Updated

File: contracts/src/v0.8/shared/test/access/AuthorizedCallers.t.sol (hunk 28 above)
Comment (RayXpub) [reply contracts/src/v0.8/shared/test/access/AuthorizedCallers.t.sol]: Updated

File: contracts/src/v0.8/shared/access/AuthorizedCallers.sol (hunk 27 above)
Comment (RayXpub) [reply contracts/src/v0.8/shared/access/AuthorizedCallers.sol]: Added

File: contracts/src/v0.8/shared/access/AuthorizedCallers.sol (hunk 26 above)
Comment (RayXpub) [reply contracts/src/v0.8/shared/access/AuthorizedCallers.sol]: Added

File: contracts/src/v0.8/shared/access/AuthorizedCallers.sol (hunk 25 above)
Comment (RayXpub) [reply contracts/src/v0.8/shared/access/AuthorizedCallers.sol]: Fixed

File: contracts/gas-snapshots/functions.gas-snapshot (hunk 23 above)
Comment (RayXpub) [reply contracts/gas-snapshots/functions.gas-snapshot]: I am not sure but diffs are gone after updating gas snapshots.

File: contracts/gas-snapshots/shared.gas-snapshot (hunk 24 above)
Comment (RayXpub) [reply contracts/gas-snapshots/shared.gas-snapshot:9]: Opened a new PR on the chainlink repo: https://github.com/smartcontractkit/chainlink/pull/13625

File: contracts/src/v0.8/ccip/NonceManager.sol (hunk 14 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/NonceManager.sol]: I was thinking of not adding more changes to this PR but I agree since we are already modifying the arl it makes sense to do it now. I updated the contract.

**Review by elatoskinas — APPROVED** [review]

File: contracts/src/v0.8/ccip/NonceManager.sol (hunk 29 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/NonceManager.sol]: `getIncrementedOutboundNonce`?

Comment (RayXpub) [reply contracts/src/v0.8/ccip/NonceManager.sol]: Updated

File: contracts/src/v0.8/ccip/NonceManager.sol [hunk 32]
```diff
@@ -0,0 +1,102 @@
...
+  /// @dev Struct that contains the previous on/off ramp addresses
+  // TODO: add prevOffRamp
+  struct PreviousRamps {
+    address prevOnRamp; // Previous onRamp
+  }
+
+  /// @dev Struct that contains the chain selector and the previous on/off ramps, same as PreviousRamps but with the chain selector
+  /// so that an array of these can be passed to the applyPreviousRampsUpdates function
+  struct PreviousRampsArgs {
+    uint64 remotChainSelector; // Chain selector
```
Comment (connorwstein) [inline contracts/src/v0.8/ccip/NonceManager.sol]: typo

File: contracts/src/v0.8/ccip/NonceManager.sol [hunk 33]
```diff
@@ -0,0 +1,102 @@
...
+  /// @dev Struct that contains a sender's outbound and inbound nonces
+  struct Nonce {
+    uint64 outboundNonce; // ──╮ Outbound nonce used by the onramp
+    uint64 inboundNonce; // ───╯ Inbound nonce used by the offramp
+  }
+
+  /// @dev previous ramps
+  mapping(uint64 chainSelector => PreviousRamps previousRamps) private s_previousRamps;
+  /// @dev The current nonces per sender used on the on/off ramps
+  mapping(uint64 remoteChainSelector => mapping(bytes sender => Nonce nonce)) private s_nonces;
```
Comment (connorwstein) [inline contracts/src/v0.8/ccip/NonceManager.sol]: hmm doesn't this grouping seem a bit hard to follow? For say s_nonces[optimism][addressA] = {Outbound: 10, Inbound: 11}, the outbound nonce 10 interprets optimism as a destination whereas inbound 11 interprets optimism as a source. An alternative would just be parallel maps which I think wouldn't cost more in the hot path (might even cost less as we avoid the encoding in the outbound direction) and be a bit more easy to follow:
```
s_outboundNonces[uint64 destChain][address sender]{uint64 outboundNonce}

s_inboundNonces[uint64 sourceChain][bytes sender]{uint64 inboundNonce}
```

Comment (RayXpub) [reply contracts/src/v0.8/ccip/NonceManager.sol]: Yes I originally used 2 mappings and then switched to the struct as it felt cleaner but I can see it being confusing, I'll update.

File: contracts/src/v0.8/ccip/NonceManager.sol (hunk 32 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/NonceManager.sol]: Fixed

File: contracts/src/v0.8/ccip/NonceManager.sol (hunk 33 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/NonceManager.sol]: Updated

File: contracts/src/v0.8/ccip/NonceManager.sol (hunk 30 above)
Comment (RayXpub) [reply contracts/src/v0.8/ccip/NonceManager.sol]: Ended up updating it with https://github.com/smartcontractkit/ccip/pull/985#discussion_r1649142047

**Review by connorwstein — APPROVED** [review]
